packages/firestore/src/local/local_store.ts

```diff
@@ -862,9 +882,16 @@ export class LocalStore {
   ): PersistencePromise<DocumentKeySet> {
     let promiseChain = PersistencePromise.resolve();
     for (const batchResult of batchResults) {
-      promiseChain = promiseChain.next(() =>
-        this.applyWriteToRemoteDocuments(txn, batchResult, documentBuffer)
-      );
+      promiseChain = promiseChain
+        .next(() =>
+          this.applyWriteToRemoteDocuments(txn, batchResult, documentBuffer)
+        )
+        .next(() =>
+          this.sharedClientState.trackMutationResult(
+            batchResult.batch.batchId,
+            'acknowledged'
```

Comment:
So we're notifying LocalStorage before the transaction has been committed, which seems like it could cause race condition issues...  This is perhaps a good reason for LocalStore not to have access to SharedClientState.  In general it won't be safe to do anything with SharedClientState in the middle of transactions (and most LocalStore methods are transactional).

Reply:
Ahh, good call!  I forgot that IndexedDb transactions are exclusive and we are inside of one.  Thanks for adding the comment though.